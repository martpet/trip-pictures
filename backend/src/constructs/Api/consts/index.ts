export * from './authorizationHeader';
export * from './lambdaEdgeViewerEvent';
