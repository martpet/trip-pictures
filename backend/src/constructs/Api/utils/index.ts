export * from './cookieName';
export * from './ddbHelpers';
export * from './edgeErrorResponse';
export * from './errorResponse';
export * from './getIdTokenPayload';
export * from './parseEventCookies';
export * from './parseLambdaEdgeEventCookies';
export * from './testHelpers';
