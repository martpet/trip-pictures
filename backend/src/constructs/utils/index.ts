export * from './createEdgeFunction';
export * from './createNodejsFunction';
export * from './createTable';
