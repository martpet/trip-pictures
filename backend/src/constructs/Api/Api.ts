import { HttpApi } from '@aws-cdk/aws-apigatewayv2-alpha';
import { Fn, NestedStack, RemovalPolicy } from 'aws-cdk-lib';
import { Distribution } from 'aws-cdk-lib/aws-cloudfront';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { ARecord, RecordTarget } from 'aws-cdk-lib/aws-route53';
import { CloudFrontTarget } from 'aws-cdk-lib/aws-route53-targets';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

import { Auth, Photos, Zone } from '~/constructs';
import { apiSubdomain, appEnvs } from '~/consts';
import { getCdkEnv } from '~/utils';

import { createDistroBehaviors } from './endpoints/createDistroBehaviors';
import { createRoutes } from './endpoints/createRoutes';

type Props = {
  zone: Zone;
  auth: Auth;
  photos: Photos;
};

export class Api extends NestedStack {
  constructor(scope: Construct, { zone, auth, photos }: Props) {
    super(scope, 'Api');

    const envName = getCdkEnv();
    const { appDomain } = appEnvs[envName];
    const apiDomain = `${apiSubdomain}.${appDomain}`;
    const api = new HttpApi(this, 'http-api');
    const origin = new HttpOrigin(Fn.select(1, Fn.split('://', api.apiEndpoint)));

    const logBucket = new Bucket(this, 'Distro-Log-Bucket', {
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });

    const distribution = new Distribution(this, 'Distro', {
      defaultBehavior: { origin },
      domainNames: [apiDomain],
      certificate: zone.certificate,
      enableLogging: true,
      logIncludesCookies: true,
      logBucket,
    });

    createRoutes({ scope, api, auth, photos });

    createDistroBehaviors({
      scope,
      api,
      auth,
      distribution,
      origin,
    });

    new ARecord(this, 'A-Record', {
      zone: zone.hostedZone,
      recordName: apiSubdomain,
      target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
    });
  }
}
