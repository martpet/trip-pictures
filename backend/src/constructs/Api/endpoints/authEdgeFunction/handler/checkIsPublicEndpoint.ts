import { isPublicEndpoint } from '~/../../shared/utils/isPublicEndpoint';
import { LambdaEdgeViewerEvent } from '~/constructs/Api/types';

export function checkIsPublicEndpoint(edgeEvent: LambdaEdgeViewerEvent) {
  const { uri, method } = edgeEvent.Records[0].cf.request;

  return isPublicEndpoint({
    path: uri,
    method,
  });
}
