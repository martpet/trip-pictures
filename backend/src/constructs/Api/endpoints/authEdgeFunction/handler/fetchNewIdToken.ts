import fetch from 'node-fetch';

import { FetchNewIdTokenResponse } from '~/constructs/Api/types';
import { authPaths } from '~/consts';

import { updateSession } from './updateSession';

type FetchNewIdTokenProps = {
  refreshToken: string;
  sessionId: string;
  clientId: string;
};

export const fetchNewIdToken = async ({
  refreshToken,
  sessionId,
  clientId,
}: FetchNewIdTokenProps) => {
  const { authDomain } = globalAuthEdgeFunctionProps;

  console.info('Fetching new id token...');

  const response = await fetch(`https://${authDomain}${authPaths.token}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: new URLSearchParams({
      grant_type: 'refresh_token',
      client_id: clientId,
      refresh_token: refreshToken,
    }),
  });

  const data = (await response.json()) as FetchNewIdTokenResponse;

  if ('error' in data) {
    throw new Error(data.error);
  }

  await updateSession({ sessionId, idToken: data.id_token });

  console.info('New id token fetched and saved');

  return data.id_token;
};
