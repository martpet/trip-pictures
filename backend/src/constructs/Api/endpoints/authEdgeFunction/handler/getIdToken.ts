import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, GetCommand } from '@aws-sdk/lib-dynamodb';

import { getIdTokenPayload } from '~/constructs/Api/utils/getIdTokenPayload';
import { refreshTokenExpiredErrorMessage, region, sessionsTableOptions } from '~/consts';
import { SessionsTableItem } from '~/types';

import { fetchNewIdToken } from './fetchNewIdToken';

const ddbClient = new DynamoDBClient({ region });
const ddbDocClient = DynamoDBDocumentClient.from(ddbClient);

export const getIdToken = async (sessionId: string) => {
  const getCommand = new GetCommand({
    TableName: sessionsTableOptions.tableName,
    Key: { [sessionsTableOptions.partitionKey.name]: sessionId },
  });

  const { Item } = await ddbDocClient.send(getCommand);
  const sessionsItem = Item as SessionsTableItem | undefined;

  if (!sessionsItem) {
    throw new Error(`could not find session with id "${sessionId}"`);
  }

  const { idToken, refreshToken, refreshTokenExpires } = sessionsItem;
  const { exp: idTokenExpires, aud: clientId } = await getIdTokenPayload(idToken);

  if ((idTokenExpires - 5) * 1000 > Date.now()) {
    return idToken;
  }

  if (refreshTokenExpires * 1000 > Date.now()) {
    return fetchNewIdToken({ refreshToken, sessionId, clientId });
  }

  throw new Error(refreshTokenExpiredErrorMessage);
};
