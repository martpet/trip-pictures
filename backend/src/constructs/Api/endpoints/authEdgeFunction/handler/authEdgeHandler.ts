import clone from 'clone';
import { StatusCodes } from 'http-status-codes';

import { LambdaEdgeViewerRequestHandler } from '~/constructs/Api/types';
import { edgeErrorResponse } from '~/constructs/Api/utils/edgeErrorResponse';
import { parseLambdaEdgeEventCookies } from '~/constructs/Api/utils/parseLambdaEdgeEventCookies';

import { checkIsPublicEndpoint } from './checkIsPublicEndpoint';
import { getIdToken } from './getIdToken';

export const handler: LambdaEdgeViewerRequestHandler = async (event) => {
  try {
    const { request } = event.Records[0].cf;
    delete request.headers.authorization;

    if (request.method === 'OPTIONS') {
      return { status: StatusCodes.NO_CONTENT };
    }

    if (checkIsPublicEndpoint(event)) {
      return request;
    }

    const { sessionId } = parseLambdaEdgeEventCookies(event);

    if (!sessionId) {
      throw new Error('invalid or missing `sessionId` cookie');
    }

    const requestClone = clone(request);
    requestClone.headers.authorization = [{ value: await getIdToken(sessionId) }];
    return requestClone;
  } catch (error) {
    return edgeErrorResponse('eWUjcKPBi3', {
      statusCode: StatusCodes.UNAUTHORIZED,
      error,
    });
  }
};
