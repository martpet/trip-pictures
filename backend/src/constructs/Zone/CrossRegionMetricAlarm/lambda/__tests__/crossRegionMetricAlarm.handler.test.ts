import {
  CloudWatchClient,
  DeleteAlarmsCommand,
  DeleteAlarmsCommandOutput,
  PutMetricAlarmCommand,
} from '@aws-sdk/client-cloudwatch';
import { mockClient } from 'aws-sdk-client-mock';

import { itResolves, itSendsAwsCommand } from '~/constructs/Api/utils';

import { handler } from '../crossRegionMetricAlarm.handler';

const cloudWatchMock = mockClient(CloudWatchClient);

const args = [
  {
    ResourceProperties: {
      region: 'dummyRegion',
      putMetricAlarmInput: {
        AlarmName: 'dummyAlarmName',
      },
    },
  },
] as unknown as Parameters<typeof handler>;

beforeEach(() => {
  cloudWatchMock.reset();

  cloudWatchMock
    .on(DeleteAlarmsCommand)
    .resolves('dummyDeleteAlarmsCommandOutput' as unknown as DeleteAlarmsCommandOutput);
});

describe('crossRegionMetricAlarm.handler', () => {
  itSendsAwsCommand(PutMetricAlarmCommand, cloudWatchMock, handler, args);
  itResolves(handler, args);

  describe('when "RequestType" is "Delete"', () => {
    const argsClone = structuredClone(args);
    argsClone[0].RequestType = 'Delete';
    itSendsAwsCommand(DeleteAlarmsCommand, cloudWatchMock, handler, argsClone);
    itResolves(handler, argsClone);
  });
});
