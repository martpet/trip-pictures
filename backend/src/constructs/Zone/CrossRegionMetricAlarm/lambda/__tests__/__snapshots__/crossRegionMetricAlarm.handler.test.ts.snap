// Vitest Snapshot v1

exports[`crossRegionMetricAlarm.handler > resolves with a correct value 1`] = `
{
  "PhysicalResourceId": "dummyAlarmName",
}
`;

exports[`crossRegionMetricAlarm.handler > sends "PutMetricAlarmCommand" from "CloudWatchClient" with correct args 1`] = `
{
  "AlarmName": "dummyAlarmName",
}
`;

exports[`crossRegionMetricAlarm.handler > when "RequestType" is "Delete" > resolves with a correct value 1`] = `"dummyDeleteAlarmsCommandOutput"`;

exports[`crossRegionMetricAlarm.handler > when "RequestType" is "Delete" > sends "DeleteAlarmsCommand" from "CloudWatchClient" with correct args 1`] = `
{
  "AlarmNames": [
    "dummyAlarmName",
  ],
}
`;
